### ISSUE TITLE
support multiple pointer devices

### ISSUE THREAD

**Issue by karroffel** [bevyengine/bevy#15]
The current `Input<MouseButton>` interface is very nice but implies the existence of a single mouse input device.

In future maybe multiple mice or touch inputs want to be supported (touch having a "backwards compatible" interface with `MouseButton` maybe?), for that a mechanism to select/retrieve the input device ID would be needed.

**Comment by jamadazi** [comment]
Now that bevy has an API for touch input (and since mice and touch are different things anyway), can we adjust this issue (change the title, etc) to be about supporting multiple mice / pointer devices, specifically?

**Comment by bushrat011899** [comment]
This feels like a surprisingly straight-forward problem to be the oldest open issue on Bevy.

**Comment by Lege19** [comment]
I'm looking into working on this as my first contribution, and I'd like some affirmation that the way I have a sensible plan:
- `MultiMouseButtonInput`: event struct containing a `MouseButtonInput` and a `winit::event::DeviceId`
- `MultiMouseMotion`: event struct containing a `MouseMotion` and a `winit::event::DeviceId`
- `MultiMouseWheel`: event struct containing a `MouseWheel` and a `winit::event::DeviceId`
- `MultiMouseButton`: struct containing a `MouseButton` and a `winit::event::DeviceId`
- `multi_mouse_button_input_system`: the equivalent of `mouse_button_input_system`, but for `ButtonInput<MultiMouseButton>` (I'd separate these into different systems because they do not require access to any of the same resources or event readers and so they could run in parallel)
- Update `<InputPlugin as Plugin>::build` to include `.init_event::<...>` for the above events
- Update `<InputPlugin as Plugin>::build` to include `.init_resource::<ButtonInput<MultiMouseButton>>`
- Update `<InputPlugin as Plugin>::build` to include `.add_systems(PreUpdate, multi_mouse_button_input_system.in_set(InputSystem))`
- Update `<InputPlugin as Plugin>::build` to include `.register_type::<MultiMouseButtonInput>`
- Add variants to `bevy_window::event::WindowEvent` for the above events
- Add `bevy_winit::converters::convert_multi_mouse_button(mouse_button: winit::event::MouseButton) -> MultiMouseButton`
- Update `bevy_winit::state::window_event` to emit `MultiMouseButtonInput` and `MultiMouseWheel` in addition to the existing mouse events
- Update `bevy_winit::state::device_event` to emit `MultiMouseMotion` as well as `MouseMotion`

I think I'm *not* going to try and update bevy's pointer API's to support multiple pointers associated with different pointer devices, since (I think) those are supposed to imitate the underlying OS's pointer behavior, and most (all?) OS's *don't* this.

**Comment by NthTensor** [comment]
Hey, just so you're aware, winit (our windowing lib) is moving to a pointer abstraction in its next release. So I think we are probably just going to switch to a unified pointer api (and possibly scrap a big portion of the existing bevy_input) instead of making additions to what we already have:

**Comment by Lege19** [comment]
I was not aware, thanks for saving me the trouble, I'll look for another issue that seems doable.

**Comment by NthTensor** [comment]
I'm going to close this, since we've had support for multiple pointers with the picking api for a while now.
